File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.cpp
/*
    0 -> 1
    1 -> 1

    0 -> 1
    1 -> 1
    3 -> 1
    6 -> 1 
*/
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size(), count = 0, sum = 0;
        unordered_map<int, int> prefix; prefix[0] = 1;
        for(int i = 0; i < n; i++){
            sum += nums[i];
            if(prefix[sum-k] > 0){
                count += prefix[sum-k];
            }
            prefix[sum]++;
        }
        return count;
    }
};

File: 0560-subarray-sum-equals-k/0560-subarray-sum-equals-k.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        mp = {0: 1}
        count, prefix = 0, 0

        for num in nums:
            prefix += num
            dif = prefix-k;

            if dif in mp:
                count += mp[dif]
            
            mp[prefix] = mp.get(prefix, 0)+1
        
        return count
        